gc/2022/GC-08-23.md

```diff
@@ -33,3 +33,189 @@ Installation is required, see the calendar invite.
+
+DG: The select, and br_table examples make sense, but the larger picture of exactly what problems we would run into is still hard for me to grasp. 
+
+BT: Look at the func.bind example, Andreas had an example of how this would work, without annotations we would have to do this operation, I’m not sure what the complexity would be it would have made it easier to have annoptations on func.bind
+
+JK: We can always require annotations in places where annotations are required, for example we could say that the hypothetical struct.get is valid if and only if the type of the thing on the value type is uniquely defined. If that’s not the case, a new instruction has to be used, or an explicit ref.cast or something else would need to be introduced
+
+AR: Highlighted the problem again, that’s what doesn’t work. The condition under which.. You would need to be able to decide whether the type is principled or not, and deciding that as complex as as deriving all tht types that are possible. You have a decision process that is arbitrarily expensive
+
+JK: To validate the annotation, which you have to do in an engine, you do have to check the type on the stack, for example, you have to check is there even a struct to do struct.get on that, and if you can tell easily enough whether you have a struct at all, then presumably you can always tell whether that has a type parameter or not 
+
+AR: It’s more complicated than that. In general, do you have a principled type or not? Is there a most generic type that covers all possibilities. If you don’t have that then you need back checking in general. This is a thing you don’t want to do. To even decide whether your type is principled, that’s not something you can do without backtracking either. You have to explore all the search space where you could have decided otherwise. I have seen systems that track the principality of the types used, but that is very research,  I don’t know how well it generalizes. In general, the assumption that you can make this distinction doesn’t work
+
+BT: In practise what would happen is that the annotation less instruction become less and less useful over time, because you add features that can’t use them, and then you have to come up with rules for when  they can be used and not and they can be very subtle
+
+AR: Because you’re very likely to use properties like substitutibility, if you have very adhoc restrictionson what types are allowed somewhere, one essential rule of subtyping is that whenever you can use a type, you can always use a subtype, if you break the property, you’ve broken the entire type system, so thes  rules have to work in a way that you don’t break the property, I’m not sure that it’s possible
```

Comment:
```suggestion
AR: Because you’re very likely to lose properties like substitutability, if you have very adhoc restrictionson what types are allowed somewhere. One essential principle of subtyping is that whenever you can use a type, you can always use a subtype. If you break the property, you’ve broken the entire type system, so these rules have to work in a way that you don’t break the property, I’m not sure that is possible in general
```